CHANGELOG.md

```diff
@@ -9,6 +9,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- `libcnb`:
+    - Low-level layer operations are now exposed via `BuildContext`. In most cases, users do not need to use them. They
+      serve as an escape hatch when the higher-level layers API doesn't support a certain
+      use-case. ([#786](https://github.com/heroku/libcnb.rs/pull/786))
```

Comment:
That is correct. See the PR description: 

> Why do we need this when we're aiming for a new high-level API?
>
> First of all, that new high-level API needs access to these operations. They were originally written as helpers for the Layer trait API and and needed some adjustments to make them usable in other contexts. I mainly improved error handing and naming. This work needed to be done in any case and this PR front loads these changes in a separate PR for easier review. :)
>
> Secondly, regardless of how great the new high-level API will be, there will always be fringe use-cases where low-level building blocks are needed. Right now, users of libcnb.rs are out of luck if they ever hit these uses-cases. After this PR, we have an escape hatch when a higher-level abstraction (either the current or new one) doesn't work.

Reply:
Got it. I'm onboard, though with some caveats.

Short: I think we should error when `LayerType`-s are not configured for functions: `replace_layer_exec_d_programs`, `replace_layer_sboms` and `EnvLayer::write_to_layer_dir`

Long:
----
Rather than a minimum API, I'm interested in a "minimum safe API". My major fear is that something is written without a `LayerType` in which case it would look like it's having some effect, but is silently having no effect. Which, if we make it ergonomic and possible to do, it will happen.

## What I like in this PR:

- `read_layer` any read operation won't have this problem 👍🏻
- `write_layer_metadata` by definition won't have this problem 👍🏻
- `delete_layer` listed for completeness 👍🏻

## What I don't like

- `replace_layer_exec_d_programs` 
- `replace_layer_sboms`
- `EnvLayer::write_to_layer_dir`

These all allow writing values to layers that may or may not do what the programmer expects.

## Options

- Explicitly checking that LayerType data is written to the layer on each of these API calls and returning an error if it's not. If a programmer doesn't care, they can catch and disregard the error.
- A "kitchen sink" read/write minimal API. One that accepts every possible input to a layer to force the programmer to consider the LayerType. This is more-or-less what I was advocating for before, what you're calling a "higher level" API.
- Requiring each of these functions to accept a LayerType

## Evaluation

I think the most "rust" way to handle this concern would be checking and erroring if the LayerType data is not set. It won't fail a compilation when the programmer makes a mistake, but at least it will be caught in an integration test instead of failing silently and in unexpected ways. It's a little more than "minimal" API, but I think it's within "minimal safe API."

## Suggestion

Check correct LayerType information is present and error if missing for:

- `replace_layer_exec_d_programs` 
- `replace_layer_sboms`
- `EnvLayer::write_to_layer_dir`

While `EnvLayer::write_to_layer_dir` is not in your PR, it already returns a Result and I think we should take the opportunity to standardize.

I would go a step further for ergonomics/consistency and say we should have an api `fn replace_layer_envs` on context that takes an `EnvLayer` so programmers don't miss this functionality (like I did).

Reply:
Since these have been ripped out. I'll dismiss this conversation.